### PR TITLE
Support set_gsm_signal

### PR DESCRIPTION
**Pull request by ki4070ma**
For https://github.com/appium/python-client/issues/353

### CODE REVIEW

File: appium/webdriver/extensions/gsm.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+class Gsm(webdriver.Remote):
+
+    (
+        NONE_OR_UNKNOWN,
+        POOR,
+        MODERATE,
+        GOOD,
+        GREAT
+    ) = list(map(str, range(5)))  # TODO: Need to check where int 0 will be dropped as arg
```
Comment (ki4070ma) [inline appium/webdriver/extensions/gsm.py]: Here is log when int 0, not "0", is used.
4 is used instead of 0.

```
[HTTP] --> POST /wd/hub/session/806d31f9-84d0-4bd4-8264-cd6af8dd7ebb/appium/device/gsm_signal
(★) [HTTP] {"signalStrength":0,"sessionId":"806d31f9-84d0-4bd4-8264-cd6af8dd7ebb"}
[W3C (806d31f9)] Calling AppiumDriver.gsmSignal() with args: [null,"806d31f9-84d0-4bd4-8264-cd6af8dd7ebb"]
[ADB] gsm signal-profile  changes the reported strength on next (15s) update.
[ADB] Getting connected emulators
[ADB] Getting connected devices...
[ADB] 1 device(s) connected
[ADB] 1 emulator(s) connected
(★)[ADB] Running '/Users/atsushimori/Library/Android/sdk/platform-tools/adb -P 5037 -s emulator-5554 emu gsm signal-profile 4'
```

Comment (KazuCocoa) [reply appium/webdriver/extensions/gsm.py]: the 4 is the default value
https://github.com/appium/appium-adb/blob/527c8b417d7b62717ececfdf068309b0ee4dbd8f/lib/tools/adb-emu-commands.js#L165

File: appium/webdriver/extensions/gsm.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    def set_gsm_signal(self, strength):
+        """Set GSM signal strength (Emulator only)
+
+        :Args:
+         - strength: Signal strength. Can be set Gsm.NONE_OR_UNKNOWN/POOR/MODERATE/GOOD/GREAT
+
+        :Usage:
+            self.driver.set_gsm_signal(Gsm.GOOD)
+        """
+        self.execute(Command.SET_GSM_SIGNAL, {'signalStrength': strength})
```
Comment (KazuCocoa) [inline appium/webdriver/extensions/gsm.py]: Which Appium version did you run on your local?
We should send two params here because of https://github.com/appium/appium-base-driver/pull/308

```
{'signalStrength': strength,'signalStrengh': strength})
```

File: appium/webdriver/extensions/gsm.py (hunk 1 above)
Comment (ki4070ma) [reply appium/webdriver/extensions/gsm.py]: Yes, I've checked.
It's problem that 4(Gsm.GREAT) is set on the device when ```self.driver.set_gsm_signal(0)```(Gsm.NONE_OR_UNKNOWN) is called. But ```self.driver.set_gsm_signal('0')``` works.

I'll look into appium-adb codes.

File: appium/webdriver/extensions/gsm.py (hunk 2 above)
Comment (ki4070ma) [reply appium/webdriver/extensions/gsm.py]: > Which Appium version did you run on your local?

v1.12.1

Thanks for your notice, I'll support both params.

Comment (ki4070ma) [reply appium/webdriver/extensions/gsm.py]: @KazuCocoa 

> We should send two params here because of appium/appium-base-driver#308

Let me confirm, this reason is to cover old appium case, right?
(If so, I'll check another client behavior like java.)

* appium v1.10 (accept only "signalStrengh")

I'm having below error log when sending ```{'signalStrength': strength,'signalStrengh': strength})```

```
➜  mobile git:(master) ✗ python SampleAndroidTest.py
test_01_test_name (__main__.SimpleTest) ... 0
ERROR

======================================================================
ERROR: test_01_test_name (__main__.SimpleTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "SampleAndroidTest.py", line 42, in test_01_test_name
    self.driver.set_gsm_signal(Gsm.POOR)
  File "/Users/atsushimori/.pyenv/versions/3.7.1/lib/python3.7/site-packages/appium/webdriver/extensions/gsm.py", line 38, in set_gsm_signal
    self.execute(Command.SET_GSM_SIGNAL, {'signalStrength': strength, 'signalStrengh': strength})
  File "/Users/atsushimori/.pyenv/versions/3.7.1/lib/python3.7/site-packages/selenium/webdriver/remote/webdriver.py", line 321, in execute
    self.error_handler.check_response(response)
  File "/Users/atsushimori/.pyenv/versions/3.7.1/lib/python3.7/site-packages/appium/webdriver/errorhandler.py", line 29, in check_response
    raise wde
  File "/Users/atsushimori/.pyenv/versions/3.7.1/lib/python3.7/site-packages/appium/webdriver/errorhandler.py", line 24, in check_response
    super(MobileErrorHandler, self).check_response(response)
  File "/Users/atsushimori/.pyenv/versions/3.7.1/lib/python3.7/site-packages/selenium/webdriver/remote/errorhandler.py", line 242, in check_response
    raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.InvalidArgumentException: Message: Parameters were incorrect. We wanted {"required":["signalStrengh"]} and you sent ["signalStrength","signalStrengh","sessionId"]


----------------------------------------------------------------------
Ran 1 test in 22.366s

FAILED (errors=1)
```

File: appium/webdriver/extensions/gsm.py (hunk 1 above)
Comment (KazuCocoa) [reply appium/webdriver/extensions/gsm.py]: Your `[W3C (806d31f9)] Calling AppiumDriver.gsmSignal() with args: [null,"806d31f9-84d0-4bd4-8264-cd6af8dd7ebb"]`  should be like below when you send the request to server.
`null` is the cause of using `4` which is the default value.

```
[debug] [W3C (0df92b78)] Calling AppiumDriver.gsmSignal() with args: [0,"0df92b78-87ce-441e-93c7-2baf70eef29d"]
```

Comment (ki4070ma) [reply appium/webdriver/extensions/gsm.py]: Thanks. Hum, I'll keep current implementation and remove my comments.

string "0" works.

```
[HTTP] --> POST /wd/hub/session/604c1c76-bb99-4e47-b998-7ba2590a9fa6/appium/device/gsm_signal
(★)[HTTP] {"signalStrengh":"0","sessionId":"604c1c76-bb99-4e47-b998-7ba2590a9fa6"}
[W3C (604c1c76)] Calling AppiumDriver.gsmSignal() with args: ["0","604c1c76-bb99-4e47-b998-7ba2590a9fa6"]
[ADB] gsm signal-profile  changes the reported strength on next (15s) update.
[ADB] Getting connected emulators
[ADB] Getting connected devices...
[ADB] 1 device(s) connected
[ADB] 1 emulator(s) connected
(★)[ADB] Running '/Users/atsushimori/Library/Android/sdk/platform-tools/adb -P 5037 -s emulator-5554 emu gsm signal-profile 0'
```

File: appium/webdriver/extensions/gsm.py (hunk 2 above)
Comment (KazuCocoa) [reply appium/webdriver/extensions/gsm.py]: ah, yeah. The current way looks good.
Thanks for it.

File: appium/webdriver/extensions/gsm.py (hunk 1 above)
Comment (KazuCocoa) [reply appium/webdriver/extensions/gsm.py]: I wonder Python client has an issue to filter `0`.
Could you add a unit test as the zero value?

Mine was by Ruby client. If JSON request has `0`, Appium should work.

Comment (ki4070ma) [reply appium/webdriver/extensions/gsm.py]: My guess is like below.
I'll look into codes more.

```
➜  mobile git:(master) ✗ python
Python 3.7.1 (default, Dec 29 2018, 23:35:51) 
>>> if 0:
...   print('hoge')
... 
>>> if 1:
...   print('hoge')
... 
hoge
```

Comment (KazuCocoa) [reply appium/webdriver/extensions/gsm.py]: so far, https://github.com/appium/python-client/pull/357/files#r279606290 is needed. I ensured.
Then, `0` should work

File: appium/webdriver/extensions/gsm.py (hunk 2 above)
Comment (ki4070ma) [reply appium/webdriver/extensions/gsm.py]: Changed to ```{'signalStrength': strength,'signalStrengh': strength})``` for https://github.com/appium/python-client/pull/357#discussion_r279664482

File: appium/webdriver/extensions/gsm.py (hunk 1 above)
Comment (ki4070ma) [reply appium/webdriver/extensions/gsm.py]: Thanks for your info, it works.

File: appium/webdriver/extensions/gsm.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    def set_gsm_signal(self, strength):
+        """Set GSM signal strength (Emulator only)
+
+        :Args:
+         - strength: Signal strength. Can be set Gsm.NONE_OR_UNKNOWN/POOR/MODERATE/GOOD/GREAT
+
+        :Usage:
+            self.driver.set_gsm_signal(Gsm.GOOD)
+        """
+        self.execute(Command.SET_GSM_SIGNAL, {'signalStrength': strength, 'signalStrengh': strength})
```
Comment (KazuCocoa) [inline appium/webdriver/extensions/gsm.py]: What about returning an error message if a user selects out of https://github.com/appium/python-client/pull/357/files#diff-bb5916ee1320d1333f6c9ba19cf04705R22 ?
```
if strength not in [NONE_OR_UNKNOWN, .....]:
    return 'error message'
```

Comment (ki4070ma) [reply appium/webdriver/extensions/gsm.py]: Thanks, seems good.
If throwing exception is too enough, please let me know.

File: appium/webdriver/extensions/gsm.py [hunk 4]
```diff
@@ -32,14 +32,9 @@ def set_gsm_signal(self, strength):
         :Args:
          - strength: Signal strength. Can be set Gsm.NONE_OR_UNKNOWN/POOR/MODERATE/GOOD/GREAT
 
-        :Raises:
-        - TypeError - Raises a TypeError if the argument is out of range
-
         :Usage:
             self.driver.set_gsm_signal(Gsm.GOOD)
         """
-        if strength not in [self.NONE_OR_UNKNOWN, self.POOR, self.MODERATE, self.GOOD, self.GREAT]:
```
Comment (KazuCocoa) [inline appium/webdriver/extensions/gsm.py]: Here is pretty good to show users which args can use for this method.
(Here is client-side logic matter)

Comment (ki4070ma) [reply appium/webdriver/extensions/gsm.py]: Added warning log instead of throwing exception.
Here is behavior.

```
➜  mobile git:(master) ✗ python SampleAndroidTest.py
test_01_test_name (__main__.SimpleTest) ... 100 is out of range. Use the value like Gsm.GOOD for signal strength.
ERROR

======================================================================
ERROR: test_01_test_name (__main__.SimpleTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "SampleAndroidTest.py", line 48, in test_01_test_name
    self.driver.set_gsm_signal(100)
  File "/Users/atsushimori/.pyenv/versions/3.7.1/lib/python3.7/site-packages/appium/webdriver/extensions/gsm.py", line 41, in set_gsm_signal
    self.execute(Command.SET_GSM_SIGNAL, {'signalStrength': strength, 'signalStrengh': strength})
  File "/Users/atsushimori/.pyenv/versions/3.7.1/lib/python3.7/site-packages/selenium/webdriver/remote/webdriver.py", line 321, in execute
    self.error_handler.check_response(response)
  File "/Users/atsushimori/.pyenv/versions/3.7.1/lib/python3.7/site-packages/appium/webdriver/errorhandler.py", line 29, in check_response
    raise wde
  File "/Users/atsushimori/.pyenv/versions/3.7.1/lib/python3.7/site-packages/appium/webdriver/errorhandler.py", line 24, in check_response
    super(MobileErrorHandler, self).check_response(response)
  File "/Users/atsushimori/.pyenv/versions/3.7.1/lib/python3.7/site-packages/selenium/webdriver/remote/errorhandler.py", line 242, in check_response
    raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.WebDriverException: Message: An unknown server-side error occurred while processing the command. Original error: Invalid signal strength param 100. Supported values: 0,1,2,3,4


----------------------------------------------------------------------
Ran 1 test in 20.444s

FAILED (errors=1)
```

Comment (KazuCocoa) [reply appium/webdriver/extensions/gsm.py]: what about something like `{name for name, value in vars(self).items() if name.isupper()}` to show the consts instead of the number?

Comment (ki4070ma) [reply appium/webdriver/extensions/gsm.py]: I misunderstood for https://github.com/appium/python-client/pull/357/commits/1ca58f4e7895b870c4aedd8f707fe8e5a1d0e65a

> selenium.common.exceptions.WebDriverException: Message: An unknown server-side error occurred while processing the command. Original error: Invalid signal strength param 100. Supported values: 0,1,2,3,4

This log comes from https://github.com/appium/appium-adb/blob/527c8b417d7b62717ececfdf068309b0ee4dbd8f/lib/tools/adb-emu-commands.js#L168 .
So can't handle from client side.

Added log is here.
```
➜  mobile git:(master) ✗ python SampleAndroidTest.py                                                                                                           
test_01_test_name (__main__.SimpleTest) ... 100 is out of range. Use the value in ['Gsm.NONE_OR_UNKNOWN', 'Gsm.POOR', 'Gsm.MODERATE', 'Gsm.GOOD', 'Gsm.GREAT'].
ERROR
(--snip--)
```

Comment (KazuCocoa) [reply appium/webdriver/extensions/gsm.py]: Ah, okay. I misunderstood it was the warning in this pr

**Comment by ki4070ma** [comment]
Will rename gsm.py, Gsm to gsm_signal, GsmSignal in the next pr because gsm.py, Gsm can't cover GsmVoice, GsmCall with current style.
